File: python/techai_webutils/src/techai_webutils/foundation/resilience/dedup.py
"""In-memory TTL deduplicator.

The default ``Deduplicator`` backend: a process-local map of ``key -> first-seen time``.
Mirrors Go's ``foundation/resilience/dedup.Memory``. For cross-replica dedup, a shared
store (Redis/Postgres) implements the same ``Deduplicator`` Protocol.
"""

from __future__ import annotations

import asyncio
import time


class MemoryDeduplicator:
    """A process-local ``Deduplicator`` backed by a TTL map.

    A key first ``seen`` is recorded and reported not-a-duplicate; every repeat within the
    TTL is a duplicate. ``ttl_seconds`` of 0 (the default) remembers forever (never evicts).
    Not shared across replicas -- use a Redis/Postgres backend for cross-pod dedup.
    """

    def __init__(self, ttl_seconds: float = 0.0) -> None:
        """Create the store; ``ttl_seconds`` <= 0 remembers keys forever."""
        self._ttl: float = ttl_seconds
        self._seen: dict[str, float] = {}
        self._lock: asyncio.Lock = asyncio.Lock()

    async def seen(self, key: str) -> bool:
        """Return True if ``key`` is a live (un-expired) repeat; else record it and return False."""
        now = time.monotonic()
        async with self._lock:
            first = self._seen.get(key)
            if first is not None and (self._ttl <= 0 or now - first < self._ttl):
                return True
            self._seen[key] = now
            return False

File: go/foundation/resilience/dedup/dedup.go
// Package dedup provides message deduplication for exactly-once processing under
// at-least-once delivery. The default Memory backend is single-process; a shared
// store (Redis/Postgres) is the cross-replica sibling that lands when a second replica does.
package dedup

import (
	"context"
	"sync"
	"time"

	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
)

// Memory is an in-memory Deduplicator with an optional TTL. A zero TTL means keys never
// expire (bounded only by memory); a positive TTL lets a key be re-processed after it lapses.
type Memory struct {
	// seen maps a processed key to the time it was first recorded.
	seen map[string]time.Time
	// ttl is the expiry after which a key may be re-processed (0 = never expires).
	ttl time.Duration
	// mu guards seen.
	mu sync.Mutex
}

// NewMemory builds an in-memory Deduplicator. ttl of zero remembers keys forever.
func NewMemory(ttl time.Duration) *Memory {
	return &Memory{seen: make(map[string]time.Time), ttl: ttl}
}

// compile-time check: Memory satisfies the Deduplicator contract.
var _ interfaces.Deduplicator = (*Memory)(nil)

// Seen reports whether key was already processed (a duplicate to skip) and, if not,
// records it as processed. A key older than the TTL is treated as first-seen again.
func (m *Memory) Seen(_ context.Context, key string) (bool, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	now := time.Now()
	if seenAt, ok := m.seen[key]; ok && (m.ttl == 0 || now.Sub(seenAt) < m.ttl) {
		return true, nil
	}
	m.seen[key] = now
	return false, nil
}
